openslides/motions/static/js/motions/diff.js

```diff
@@ -0,0 +1,413 @@
+    };
+
+    this._getNodeContextTrace = function(node) {
+        var context = [],
+            currNode = node;
+        while (currNode) {
+            context.unshift(currNode);
+            currNode = currNode.parentNode;
+        }
+        return context;
+    };
+
+    this._insertInternalLineMarkers = function(fragment) {
+        if (fragment.querySelectorAll('OS-LINEBREAK').length > 0) {
+            // Prevent duplicate calls
+            return;
+        }
+        var lineNumbers = fragment.querySelectorAll('span.os-line-number');
+        for (var i = 0; i < lineNumbers.length; i++) {
+            var insertBefore = lineNumbers[i];
+            while (insertBefore.parentNode.nodeType != DOCUMENT_FRAGMENT_NODE && insertBefore.parentNode.childNodes[0] == insertBefore) {
+                insertBefore = insertBefore.parentNode;
+            }
+            var lineMarker = document.createElement('OS-LINEBREAK');
+            lineMarker.setAttribute('data-line-number', lineNumbers[i].getAttribute('data-line-number'));
+            lineMarker.setAttribute('class', lineNumbers[i].getAttribute('class'));
+            insertBefore.parentNode.insertBefore(lineMarker, insertBefore);
+        }
+    };
+
+  /*
+   * Returns an array with the following values:
+   * 0: the most specific DOM-node that contains both line numbers
+   * 1: the context of node1 (an array of dom-elements; 0 is the document fragment)
+   * 2: the context of node2 (an array of dom-elements; 0 is the document fragment)
+   * 3: the index of [0] in the two arrays
+   */
+    this._getCommonAncestor = function(node1, node2) {
+        var trace1 = this._getNodeContextTrace(node1),
+            trace2 = this._getNodeContextTrace(node2),
+            commonAncestor = null,
+            commonIndex = null,
+            childTrace1 = [],
+            childTrace2 = [];
+
+        for (var i = 0; i < trace1.length && i < trace2.length; i++) {
+            if (trace1[i] == trace2[i]) {
+                commonAncestor = trace1[i];
+                commonIndex = i;
+            }
+        }
+        for (i = commonIndex + 1; i < trace1.length; i++) {
+            childTrace1.push(trace1[i]);
+        }
+        for (i = commonIndex + 1; i < trace2.length; i++) {
+            childTrace2.push(trace2[i]);
+        }
+        return {
+            'commonAncestor': commonAncestor,
+            'trace1' : childTrace1,
+            'trace2' : childTrace2,
+            'index': commonIndex
+        };
+    };
+
+    this._serializeTag = function(node) {
+        if (node.nodeType == DOCUMENT_FRAGMENT_NODE) {
+            return '';
+        }
+        var html = '<' + node.nodeName;
+        for (var i = 0; i < node.attributes.length; i++) {
+          var attr = node.attributes[i];
+          html += " " + attr.name + "=\"" + attr.value + "\"";
+        }
+        html += '>';
+        return html;
+    };
+
+    this._serializeDom = function(node, stripLineNumbers) {
```

Comment:
Can you please add some comments about the return statements inside the fucntion (why early return).